### ISSUE TITLE
relation to wasmtime?

### ISSUE THREAD

**Issue by vitiral** [wasmerio/wasmer#142]
This project seems to share many of the goals of https://github.com/CraneStation/wasmtime.

Are they two experiments with the same goals, or are they fundamentally different in some way? Very supportive of both projects!

**Comment by syrusakbary** [comment]
Thanks for asking, @vitiral.

Before starting [Wasmer](https://wasmer.io/), we took a deep look into all the different alternatives for executing WebAssembly, including wasmtime (and WAVM, wasmjit, nebulet...).

After playing with [wasmtime in October](https://github.com/CraneStation/wasmtime/tree/5f11f444823a92eb78f26321ee5b3bde79c04b2b), it didn't seemed a good fit as it was impossible to run any wasm file with it and the project seemed unmaintained (as the last significant commit was a few months distant).

<p align="center"><i>wasmtime commit log</i><br />
<a href="https://github.com/CraneStation/wasmtime/graphs/commit-activity
"><img width="731" alt="screen shot 2019-02-06 at 9 25 50 am" src="https://user-images.githubusercontent.com/188257/52362809-3b97ab80-29f6-11e9-9575-c3ca71a28abd.png"></a>
</p>

So we decided to create our own runtime.

After a month working on wasmer, the creator of wasmtime decided to put more effort on it rather than supporting wasmer.
That was a sad decision, because we talked about moving from wasmtime to wasmer but it didn't happen.

Right now, however the two projects differ on their goals.

# Main differences
* Wasmer goal is to be the universal runtime for WebAssembly
* Wasmer is trying to abstract the API from the IR itself (Cranelift), so we can use LLVM for example to target other chipsets (such as RISC-V or ARM)
* Wasmer goal is to be highly pluggable, so it's easy to plug custom ABIs (imports/exports) such as [Emscripten or CloudABI](https://github.com/wasmerio/wasmer/tree/master/lib#integrations).
* Wasmer can be [easily embedded in other applications](https://medium.com/wasmer/executing-webassembly-in-your-rust-application-d5cd32e8ce46)
* Wasmer is backed by commercial use cases and applications

Other than that, we have now a full team working on it and making sure other projects can use our runtime very easily. As far as I'm aware only one person is working on wasmtime.

**Comment by vitiral** [comment]
Awesome, thanks for the breakdown. It might be worth putting a bit of this in README (at least a link to your answer). I consider my question answered :smile:

**Comment by elimisteve** [comment]
Quite a bit can change in a year and a half (when the above exchange occurred), and I thought it would only be fair to point out to others who find this thread that the core wasmtime runtime now has almost 4x the contributors as wasmer's (204 v. 55), and has 9 contributors in the last week versus 2 for wasmer.

Here's a recent, thorough blog post that compares the two runtimes and concludes they are similar in their current functionality and performance: https://wiki.alopex.li/ActuallyUsingWasm .

I'm genuinely excited about both efforts! wasmtime has some impressive companies behind it (Mozilla, Intel, Fastly, Red Hat), but if there's a startup worth betting on it would be one that's gone through YC as Wasmer has :-) .

wasmtime: https://github.com/bytecodealliance/wasmtime/pulse

wasmer: https://github.com/wasmerio/wasmer/pulse

**Comment by syrusakbary** [comment]
Thanks for commenting @elimisteve.

Let me address some of the comments:
1. Contributors. The number of contributors you are seeing for wasmtime are on almost it's majority related the Cranelift IR. This is mainly caused because they merged both projects (wasmtime and cranelift) together 😉 (a bit deceitful move, but I guess that it is what it is!)
2. Pulse on activity for Wasmer. While it looks like wasmer doesn't have a lot of recent commits, the main reason is because we are working on a lot of stuff under the hood that will be released soon (stay tuned!).
3. Performance/Speed. Wasmer is way, way ahead on this one... stay tuned for more numbers!

We really appreciate the effort that some of their developers are doing to move wasm tooling forward, which will be great for the ecosystem on the long term and super useful for their browser.

I'd like to conclude saying that the tradeoffs to make when running Wasm server-side are quite different to when running in the browser.
And for the server-side spectrum... we are the leaders, not only because we have 100x the number of installs, or because Wasmer is 50% faster than wasmtime, but because of the great community that is behind us 😊

**Comment by Rizary** [comment]
@syrusakbary stumble upon this thread when searching about wasm runtime. Regarding your statement here:

> Pulse on activity for Wasmer. While it looks like wasmer doesn't have a lot of recent commits, _the main reason is because we are working on a lot of stuff under the hood that will be released soon (stay tuned!)_.

Is there any possibility to accept community contribution to work that is "under the hood"?

**Comment by nlewycky** [comment]
> Is there any possibility to accept community contribution to work that is "under the hood"?

No need! The formerly under the hood stuff was all merged into master in #1520 at which point we've continued work in the open.

**Comment by elimisteve** [comment]
> ...Wasmer is much faster than wasmtime (50% faster in coremark)

Latest benchmarks show wasmtime significantly faster than wasmer:

https://github.com/jedisct1/webassembly-benchmarks/tree/master/2021-Q1

@syrusakbary Is libsodium not your preferred benchmark?

**Comment by syrusakbary** [comment]
> Latest benchmarks show wasmtime significantly faster than wasmer:

I think you might have misread the benchmarks.
In all the [benchmarks done in libsodium](https://github.com/jedisct1/webassembly-benchmarks/tree/master/2021-Q1), wasmer (with LLVM compiler) performs way faster than wasmtime! (and we are working to increase even more the performance in LLVM!)

The benchmarks also showed that something weird is happening when using Wasmer and the Cranelift compiler. Wasmer should be (at least) as fast as wasmtime there! (while also being much faster in startup time) We are now investigating the issue, and shall be solved quickly 🤗

**⚠️ Important**: Please note that we only recommend Cranelift for development proposes, for production we recommend using [LLVM](https://github.com/wasmerio/wasmer/tree/master/lib/compiler-llvm)!

**Comment by elimisteve** [comment]
![benchmarks](https://raw.githubusercontent.com/jedisct1/webassembly-benchmarks/master/2021-Q1/macos.png)

When I'm using wasmer on Linux, which backend is being used?

wasicc is great, btw; worked better for me than the alternatives 👍 .

**Comment by hanabi1224** [comment]
> Important: Please note that we only recommend Cranelift for development proposes, for production we recommend using LLVM

@syrusakbary @elimisteve  Very interesting topic, but it seems that LLVM backend is not available in the go package I just looked into. Below code prints false, how can I get LLVM backend working?
```go
available := wasmer.IsCompilerAvailable(wasmer.LLVM)
fmt.Println(available)
```

And just FYI, I also tried benchmarking two compilers with problems from [the programming language benchmarks game](https://benchmarksgame-team.pages.debian.net/benchmarksgame/), it shows that LLVM backend is doing better with pidigits but worse with nbody.  My results are [here](https://another-benchmarks-game.vercel.app/rust)

p.s. I use ```wasmer run --llvm <file>``` and ```wasmer run --cranelift <file>``` to target different backends, please let me know if that's not the correct way.

![image](https://user-images.githubusercontent.com/6546635/109282265-43824c80-7858-11eb-903d-daf483cf1437.png)

![image](https://user-images.githubusercontent.com/6546635/109282334-5ac13a00-7858-11eb-874a-150965feb46e.png)

**Comment by syrusakbary** [comment]
Yeah, something funny is happening on nbody/llvm. LLVM should be much faster.
We will review, thanks for bringing it up!

**Comment by rambip** [comment]
I don't know if it is the right place to ask (it's more for a duscussion, really) but how is this project related to llvm ?

What representation takes more space, webassembly or llvm-ir ?
Would that be a good idea to distribute llvm-ir with wasi symbols so that people can assemble them on their machine ?

Because primarly, the goals of wasi and llvm are quite similar: creating a standard intermediate representation between the language and the assembly run by the cpu.

Do you think I should ask wasi on theur github ?

**Comment by Hywan** [comment]
This is not the correct issue to ask such questions. Please fill another issue :-).

(And no, Wasm and LLVM IR aren't the same, to make it short).

**Comment by Hywan** [comment]
@hanabi1224 With #2161, LLVM is fast again with `nbody` (and similar float intensive computations).

**Comment by hanabi1224** [comment]
@Hywan Thanks for letting me know, and yeah I can see the improvement in my [test rest](https://programming-language-benchmarks.vercel.app/rust)
